packages/host/app/resources/room.ts

```diff
@@ -4,34 +4,335 @@ import { tracked } from '@glimmer/tracking';
+          (event.content as CardMessageContent).clientGeneratedId ?? event_id,
+          messageField as any,
+        );
+      }
+    }
+
+    for (let [id, message] of newMessages) {
+      // The `id` can either be an `eventId` or `clientGeneratedId`.
+      // For messages sent by the user, we prefer to use `clientGeneratedId`
+      // because `eventId` can change in certain scenarios,
+      // such as when resending a failed message or updating its status from sending to sent.
+      this._messageCache.set(id, message);
+    }
+  }
+
+  upsertRoomMember({
+    roomId,
+    userId,
+    displayName,
+    membership,
+    membershipDateTime,
+    membershipInitiator,
+  }: RoomMemberInterface): RoomMember {
+    let member: RoomMember | undefined;
+    member = this._memberCache.get(userId);
+    if (
+      // Create new member if it doesn't exist or if provided data is more recent
+      member?.membershipDateTime &&
+      membershipDateTime &&
+      member.membershipDateTime.getTime() > Number(membershipDateTime)
+    ) {
+      return member;
+    }
+    if (!member) {
+      let member = new RoomMember({ userId, roomId });
+      if (displayName) {
+        member.displayName = displayName;
+      }
+      if (membership) {
+        member.membership = membership;
+      }
+      if (membershipDateTime != null) {
+        member.membershipDateTime = new Date(membershipDateTime);
+      }
+      if (membershipInitiator) {
+        member.membershipInitiator = membershipInitiator;
+      }
+
+      this._memberCache.set(userId, member);
+      return member;
+    }
+    return member;
+  }
+
+  private serializedCardFromFragments(
+    eventId: string,
+  ): LooseSingleCardDocument {
+    let fragments: CardFragmentContent[] = [];
+    let currentFragment: string | undefined = eventId;
+    do {
+      let fragment = this._fragmentCache.get(currentFragment);
+      if (!fragment) {
+        throw new Error(
+          `No card fragment found in cache for event id ${eventId}`,
+        );
+      }
+      fragments.push(fragment);
+      currentFragment = fragment.data.nextFragment;
+    } while (currentFragment);
+
+    fragments.sort((a, b) => (a.data.index = b.data.index));
+    if (fragments.length !== fragments[0].data.totalParts) {
+      throw new Error(
+        `Expected to find ${fragments[0].data.totalParts} fragments for fragment of event id ${eventId} but found ${fragments.length} fragments`,
+      );
+    }
+
+    let cardDoc = JSON.parse(
+      fragments.map((f) => f.data.cardFragment).join(''),
+    ) as LooseSingleCardDocument;
+    return cardDoc;
+  }
 }
 
-export function getRoom(parent: object, roomId: () => string | undefined) {
+export function getRoom(
+  parent: object,
+  roomId: () => string | undefined,
+  events: () => any | undefined, //we need this to react to new roomAddEvent
+) {
   return RoomResource.from(parent, () => ({
     named: {
       roomId: roomId(),
+      events: events ? events() : [],
```

Comment:
suspicious why we are passing this in instead of looking it up from the matrix service

Reply:
There are 2 reactive-driven actions 
- if we change a roomId
- if we send a message

The reason this is setup this way its to react to sending new messages. When .events is set on the service ie a message is sent, it doesn't react. 

However, I have tried to remove the argument and I found that I need to call `.events` before `await getRoom`. If I dont, then sending a new message will not render in the chat. Perhaps, this is something we can pair on. I find it really hard to anticipate the asynchrony + tracked nature of the matrix setup. 

```
 private load = restartableTask(async (roomId: string) => {
    try {
      let events = this.matrixService.getRoom(named.roomId)?.events;
      console.log('events', events); // I need to access .events before getRoom await
      this.room = roomId ? await this.matrixService.getRoom(roomId) : undefined;
      console.log('room', this.room?.events);
      // if not awaited we obtain "Assertion Failed: You attempted to update _value on TrackedStorageImpl, but it had already been used previously in the same computation." Ensure that asynchronous operations are properly awaited. By using await, we can defer the execution, ensuring that the tracked property assignment (this.room) happens outside of the current synchronous computation. This avoids the reactivity conflict.
      await this.loadRoomMembers(roomId);
      await this.loadRoomMessages(roomId);

    } catch (e) {
      throw new Error(`Error loading room resource: ${e}`);
    }
  });
```

Reply:
ah yes, there is a work around for this in the lib/matrix-handlers/index.ts:

in the addRoomEvent() function:
``` ts
// duplicate events may be emitted from matrix, as well as the resolved room card might already contain this event
  if (!resolvedRoom.events.find((e) => e.event_id === eventId)) {
    resolvedRoom.events = [
      ...(resolvedRoom.events ?? []),
      event as unknown as DiscreteMatrixEvent,
    ];
  }
```

the nuance here (which is not really documented), is that by performing the array assignment `resolvedRoom.events = [... ]` we are actually triggering the rerender of the room field. The simple thing would have been to mutate the array with the event, but our reactivity in cards uses glimmer which has this same drawback in that the invalidation of the property doesn't see down into nested object mutation. This is way `TrackedArray` exists. if you make a tracked field that happens to be an array, mutating it will not cause a rerender. Same with our card fields (this is something that we plan to address with our field refactoring). But for now any room event that was added, we would actually clone the array, add the new event, and then reassign the array to the field. this does actually trigger the rerender.

So if we move to a resource, then yes, I see the dilemma now. We want that same kind of reactivity when events are added room state (in this case the state living in the room resource). Given that, i think this is ok--but lets document why we are doing that in our comments: specifically that we are passing in events within the room resource thunk so that we get reactivity for room events.

Reply:
although, actually wait--maybe that's heavy handed. I think what this means is that we'll generate a new room resource for each event. Maybe what we need is to actually have the room resource register a callback with matrix such that when new events are received, if they are events for the room, then we add the events to a tracked array within the room resource.

Reply:
I see. 

That seems a little big. Would it be wise to merge this in with a comment explaining the usage of events and the instantiation of a resource argument before optimising for listening to single event and modifying the matrix listener. 

Reason I think so is bcos, there are 2 tickets which kinda work with this
- https://linear.app/cardstack/issue/CS-6965/on-load-fetch-room-metadata-without-fetching-all-messages-from-each
- https://linear.app/cardstack/issue/CS-6986/create-an-independent-resource-on-the-panel-component-to-manage